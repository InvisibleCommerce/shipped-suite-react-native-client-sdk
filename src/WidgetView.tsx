import React, {
    forwardRef,
    ForwardRefRenderFunction,
    useImperativeHandle,
    useRef
} from 'react';
import {
    findNodeHandle,
    requireNativeComponent,
    StyleProp,
    UIManager,
    ViewStyle
} from 'react-native';

export enum ShippedSuiteType {
  Shield,
  Green,
  GreenAndShield,
}

export interface WidgetViewProps {
  style?: StyleProp<ViewStyle>;
  ref?: any;
  isMandatory?: boolean;
  isRespectServer?: boolean;
  type?: ShippedSuiteType;
  onChange: (values: any) => void;
}

export type WidgetViewMethods = {
  updateOrderValue: (amount: string) => void;
};

export interface WidgetChangeEventData {
  isSelected: boolean;
  shieldFee?: string;
  greenFee?: string;
  error?: string;
}

const RNWidgetView = requireNativeComponent<WidgetViewProps>('RNWidgetView');

const _WidgetView: ForwardRefRenderFunction<
  WidgetViewMethods,
  WidgetViewProps
> = (props, ref) => {
  const { type, isMandatory, isRespectServer, onChange, ...others } = props;
  const widgetRef = useRef<any>(null);

  const updateOrderValue = (amount: string) => {
    UIManager.dispatchViewManagerCommand(
      findNodeHandle(widgetRef.current),
      'updateOrderValue' as any,
      [amount]
    );
  };

  useImperativeHandle(
    ref,
    () => ({
      updateOrderValue,
    }),
    []
  );

  return (
    <RNWidgetView
      ref={widgetRef}
      type={type}
      isMandatory={isMandatory}
      isRespectServer={isRespectServer}
      onChange={onChange}
      {...others}
    />
  );
};

export const WidgetView = forwardRef(_WidgetView);
