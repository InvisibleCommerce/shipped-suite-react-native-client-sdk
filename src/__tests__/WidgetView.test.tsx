/**
 * @format
 */

import 'react-native';
import React from 'react';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { ShippedSuiteType, WidgetView } from '../WidgetView';

it('renders a WidgetView using Snapshots', () => {
  const component = renderer.create(
    <WidgetView
      configuration={{
        type: ShippedSuiteType.GreenAndShield,
        isInformational: true,
        isMandatory: true,
        isRespectServer: true,
      }}
      onChange={() => {}}
    />
  );
  expect(component).toMatchSnapshot();
});
