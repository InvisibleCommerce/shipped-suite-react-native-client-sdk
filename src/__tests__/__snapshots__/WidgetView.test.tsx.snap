// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a WidgetView using Snapshots 1`] = `
<RNWidgetView
  configuration={
    {
      "appearance": 0,
      "currency": "USD",
      "isInformational": true,
      "isMandatory": true,
      "isRespectServer": true,
      "type": 2,
    }
  }
  onChange={[Function]}
/>
`;
